### PR TITLE
Code Quality: Only execute actions once

### LINKED ISSUE #11737: Feature: Only open single tab if user holds down ctrl + t
### What feature or improvement do you think would benefit Files?

Holding down `ctrl` + `t` will keep opening tabs until the user releases the keys. In order to reduce the risk of crashing the app, we should only open a single tab. Depending on the context, we should do this for other shortcuts as well.

### Requirements

- Only open a single tab when holding down `ctrl` + `t`
- Only close one tab when holding down `ctrl` + `w`
- Only refresh tab once when holding down `ctrl` + `r`
- Only go back once when holding down `alt` + `left`
- Only go forward once when holding down `alt` + `right`

### Files Version

v2.4

### Windows Version

Windows 11

### Comments

_No response_

### PR DESCRIPTION
**Pull request by HorseBandit**
**Resolved / Related Issues**
- [x] Were these changes approved in an issue or discussion with the project maintainers? In order to prevent extra work, feature requests and changes to the codebase must be approved before the pull request will be reviewed. This prevents extra work for the contributors and maintainers.
   Closes #11737 

**Validation**
How did you test these changes?
- [x] Did you build the app and test your changes?
- [ ] Did you check for accessibility? You can use Accessibility Insights for this.
- [ ] Did you remove any strings from the en-us resource file?
   - [ ] Did you search the solution to see if the string is still being used? 
- [x] Did you implement any design changes to an existing feature?
   - [ ] Was this change approved?
- [x] Are there any other steps that were used to validate these changes?
   1. Open app ...
   2. hold ctrl + t and see that tabs don't continuously open
   3. press ctrl + t a few times
   4. hold ctrl +w and see that tabs don't continuously close
   5. press ctrl + w a few times
   6. refresh the same

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

**Comment by HorseBandit** [comment]
@yaira2 Ready for review. Fully implemented.

**Comment by yaira2** [comment]
@QuaintMako @Ferrario-Filippo would it be worth extending this behavior to every action? If we do this, we can handle it across the board instead of duplicating the code for each action.

**Comment by HorseBandit** [comment]
@yaira2 Thank you for the review. I'm waiting for further instruction. Is this on track to be merged to main?

**Comment by yaira2** [comment]
I'd like to wait for the others to provide their feedback.

**Comment by 0x5bfa** [comment]
redo and undo actions shouldn’t be limited, but otherwise I think all actions should be.

**Comment by yaira2** [comment]
> redo and undo actions shouldn’t be limited

I think they should, we wouldn't want someone undoing multiple operations by mistake.

**Comment by ferrariofilippo** [comment]
> would it be worth extending this behavior to every action? If we do this, we can handle it across the board instead of duplicating the code for each action.

I think we should do that, I can't think of any use-case where a user needs to execute the same action over and over. It may also prevent some bugs we are not aware of.

**Comment by yaira2** [comment]
@HorseBandit can you apply this logic to all actions? Instead of copying it to every action, I think we should build it into the underlying code.

**Comment by HorseBandit** [comment]
@yaira2 @ferrariofilippo change fully implemented for all actions. Ready for you.

File: src/Files.App/Actions/DebounceAction.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		}
+
+		public abstract string Label { get; }
+		public abstract string Description { get; }
+		public abstract RichGlyph Glyph { get; }
+		public abstract HotKey HotKey { get; }
+		public abstract HotKey SecondHotKey { get; }
+		public abstract HotKey ThirdHotKey { get; }
+		public abstract HotKey MediaHotKey { get; }
+		public abstract bool IsExecutable { get; }
```
Comment (QuaintMako) [inline src/Files.App/Actions/DebounceAction.cs:36]: These should be move above the constructor to keep the structure consistency. (See Class Structure in [the coding style](https://files.community/docs/contributing/code-style )if you're curious 😄 )

File: src/Files.App/Actions/DebounceAction.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+		public virtual bool CanExecuteNow()
+		{
+			return DateTime.Now - lastExecuted > debounceTime;
+		}
+
+		public virtual void MarkAsExecuted()
+		{
+			lastExecuted = DateTime.Now;
+		}
```
Comment (QuaintMako) [inline src/Files.App/Actions/DebounceAction.cs:27]: The name is not exactly telling of what it's doing. It should be something like `MarkLastExecutionTime`.

File: src/Files.App/Data/Commands/Manager/CommandManager.cs [hunk 3]
```diff
@@ -199,149 +199,149 @@ public CommandManager()
...
-			[CommandCodes.ClosePane] = new ClosePaneAction(),
-			[CommandCodes.OpenFileLocation] = new OpenFileLocationAction(),
-			[CommandCodes.PlayAll] = new PlayAllAction(),
-			[CommandCodes.GitFetch] = new GitFetchAction(),
-			[CommandCodes.GitInit] = new GitInitAction(),
-			[CommandCodes.GitPull] = new GitPullAction(),
-			[CommandCodes.GitPush] = new GitPushAction(),
-			[CommandCodes.GitSync] = new GitSyncAction(),
-			[CommandCodes.OpenAllTaggedItems] = new OpenAllTaggedActions(),
+			[CommandCodes.OpenHelp] = new DebouncedActionDecorator(new OpenHelpAction(), TimeSpan.FromMilliseconds(800)),
```
Comment (QuaintMako) [inline src/Files.App/Data/Commands/Manager/CommandManager.cs:202]: Seeing how often the `TimeSpan.FromMilliseconds(800)` is repeated, why not directly put it inside the constructor of `DebouncedActionDecorator` as a default value, and create a second constructor that'd take a timespan as second argument? 
It'd reduce the repetition, simplify the change of the default value and leaves the room for case to case customization.

File: src/Files.App/Actions/DebouncedActionDecorator.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		public override HotKey ThirdHotKey => _innerAction.ThirdHotKey;
+		public override HotKey MediaHotKey => _innerAction.MediaHotKey;
+		public override bool IsExecutable => _innerAction.IsExecutable;
+
+		public override async Task ExecuteAsync()
+		{
+			if (!CanExecuteNow())
+			{
+				return;
+			}
```
Comment (QuaintMako) [inline src/Files.App/Actions/DebouncedActionDecorator.cs:33]: Curly braces are not needed for one liner `if`.

File: src/Files.App/Actions/DebouncedActionDecorator.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		}
+
+		public override string Label => _innerAction.Label;
+		public override string Description => _innerAction.Description;
+		public override RichGlyph Glyph => _innerAction.Glyph;
+		public override HotKey HotKey => _innerAction.HotKey;
+		public override HotKey SecondHotKey => _innerAction.SecondHotKey;
+		public override HotKey ThirdHotKey => _innerAction.ThirdHotKey;
+		public override HotKey MediaHotKey => _innerAction.MediaHotKey;
+		public override bool IsExecutable => _innerAction.IsExecutable;
```
Comment (QuaintMako) [inline src/Files.App/Actions/DebouncedActionDecorator.cs:26]: Should be moved above the constructor

**Review by QuaintMako — COMMENTED** [review]
A few feedbacks, mostly on the format.
The functionnality looks good to me, I cannot test it for now.

**Comment by HorseBandit** [comment]
@QuaintMako @yaira2 Suggestions taken and implemented. I left one action with an optional value of 1000 milliseconds for demonstration. The others are defaulted at 800ms.

File: src/Files.App/Data/Commands/Manager/CommandManager.cs [hunk 6]
```diff
@@ -199,149 +199,149 @@ public CommandManager()
...
+			[CommandCodes.OpenNewPane] = new DebouncedActionDecorator(new OpenNewPaneAction()),
+			[CommandCodes.ClosePane] = new DebouncedActionDecorator(new ClosePaneAction()),
+			[CommandCodes.OpenFileLocation] = new DebouncedActionDecorator(new OpenFileLocationAction()),
+			[CommandCodes.PlayAll] = new DebouncedActionDecorator(new PlayAllAction()),
+			[CommandCodes.GitFetch] = new DebouncedActionDecorator(new GitFetchAction()),
+			[CommandCodes.GitInit] = new DebouncedActionDecorator(new GitInitAction()),
+			[CommandCodes.GitPull] = new DebouncedActionDecorator(new GitPullAction()),
+			[CommandCodes.GitPush] = new DebouncedActionDecorator(new GitPushAction()),
+			[CommandCodes.GitSync] = new DebouncedActionDecorator(new GitSyncAction()),
+			[CommandCodes.OpenAllTaggedItems] = new DebouncedActionDecorator(new OpenAllTaggedActions(), TimeSpan.FromMilliseconds(1000))
```
Comment (0x5bfa) [inline src/Files.App/Data/Commands/Manager/CommandManager.cs:344]: I believe we should leave as it (800 by default) is for now.

File: src/Files.App/Actions/DebounceAction.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Files.App.Actions
+{
+	public abstract class DebouncedAction : IAction
+	{
+		private DateTime lastExecuted = DateTime.MinValue;
+
+		private readonly TimeSpan debounceTime;
```
Comment (0x5bfa) [inline src/Files.App/Actions/DebounceAction.cs:13]: debounceTimeSpan

**Review by 0x5bfa — COMMENTED** [review]
being particular but needed. LGTM otherwise, thank you!

File: src/Files.App/Actions/DebouncedActionDecorator.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		/// <summary>
+		/// Initializes a new instance of the DebouncedActionDecorator class.
+		/// </summary>
+		/// <param name="innerAction">The IAction instance to be wrapped.</param>
+		/// <param name="debounceDuration">The debounce duration. Default is 800 milliseconds.</param>
+		public DebouncedActionDecorator(IAction innerAction, TimeSpan? debounceDuration = null)
+			: base(debounceDuration ?? TimeSpan.FromMilliseconds(800))
+		{
+			_innerAction = innerAction;
+		}
```
Comment (QuaintMako) [inline src/Files.App/Actions/DebouncedActionDecorator.cs:31]: How about this? It removes the possibility to have a null item, which is always a good thing.
```suggestion
		public DebouncedActionDecorator(IAction innerAction, TimeSpan debounceDuration = TimeSpan.FromMilliseconds(800))
			: base(debounceDuration)
		{
			_innerAction = innerAction;
		}
```

File: src/Files.App/Data/Commands/Manager/CommandManager.cs (hunk 6 above)
Comment (QuaintMako) [inline src/Files.App/Data/Commands/Manager/CommandManager.cs:344]: Not sure that adding an example is useful. After all, if someone needs one day to add a new command, their IDE should display the constructor with the possibility of changing the timespan. Here is just adds a "why is this command different from others?" moment when reading it.

**Comment by HorseBandit** [comment]
@0x5bfa The above steps are completed. I had to comment out a line in CommandManager. Could you check that out? It was causing a problem.

**Comment by HorseBandit** [comment]
@0x5bfa @yaira2 @QuaintMako 

Branch synced with main as of now. Ready to merge.

**Comment by HorseBandit** [comment]
@yaira2 Could I get an update on this? I think it's ready to merge. Thank you.

**Review by QuaintMako — APPROVED** [review]
LGTM.

**Review by 0x5bfa — APPROVED** [review]
LGTM

**Comment by yaira2** [comment]
@HorseBandit can you take a look at the build error?

**Comment by HorseBandit** [comment]
@yaira2 Fresh sync followed by a merge of 'main' into this branch. There are no build errors.

**Comment by hecksmosis** [comment]
I think this is a pipeline bug iirc, don't remember what we used to do to fix it

**Comment by yaira2** [comment]
@HorseBandit do the test projects run locally? It seems to be failing in the PR.

**Comment by HorseBandit** [comment]
@yaira2 Yes they do.

**Comment by yaira2** [comment]
@HorseBandit I think the best option is to duplicate the branch and create a new PR.

**Pull request closed without merging**